components/post_view/post_list/index.jsx

```diff
@@ -0,0 +1,448 @@
+            backUpPostsInChannel: PropTypes.func.isRequired,
+
+            /*
+             * Used for notifying redux store on channel sync complete
+             */
+            channelSyncCompleted: PropTypes.func.isRequired,
+
+            /*
+             * Used for adding back postIds to postsInChannel from backUpPostsInChannel
+             */
+            addPostIdsFromBackUp: PropTypes.func.isRequired,
+        }).isRequired,
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            isDoingInitialLoad: this.shouldLoadPosts(props),
+            newerPosts: {
+                loading: false,
+                allLoaded: (props.channelPostsStatus && props.channelPostsStatus.atEnd) || false,
+            },
+            olderPosts: {
+                loading: false,
+                allLoaded: this.oldestMessageLoadedInView(props),
+            },
+            lastViewedAt: props.lastViewedAt || props.member.last_viewed_at,
+        };
+    }
+
+    componentDidMount() {
+        if (this.state.isDoingInitialLoad) {
+            this.postsOnLoad(this.props.channel.id);
+        }
+        if (!this.props.channelSyncStatus && this.props.channelPostsStatus) {
+            this.syncPosts(this.props);
+        }
+    }
+
+    UNSAFE_componentWillReceiveProps(nextProps) { // eslint-disable-line camelcase
+        if (this.props.posts && nextProps.posts && (this.props.posts.length !== nextProps.posts.length)) {
+            if (this.oldestMessageLoadedInView(nextProps) !== this.state.olderPosts.allLoaded) {
+                this.setState({
+                    olderPosts: {
+                        loading: false,
+                        allLoaded: this.oldestMessageLoadedInView(nextProps),
+                    },
+                });
+            }
+        }
+        if (!this.props.socketStatus.connected && nextProps.socketStatus.connected) {
+            this.syncPosts(nextProps);
+        }
+    }
+
+    shouldLoadPosts(props) {
+        if (typeof props.channelPostsStatus === 'undefined') {
+            return true;
+        } else if ((props.match.params.messageId && !props.channelPostsStatus.atEnd) || !props.channelPostsStatus.atEnd) {
+            return true;
+        }
+
+        return false;
+    }
+
+    syncPosts = async (props) => {
+        if (props.channelPostsStatus.atEnd) {
+            await props.actions.syncChannelPosts({channelId: props.channel.id, since: props.socketStatus.lastDisconnectAt});
+        } else {
+            let data;
+            const oldestPostId = getOldestPostIdFromPosts(props.posts);
+            let newestMessageId = getNewestPostIdFromPosts(props.posts);
+            do {
+                ({data} = await props.actions.syncChannelPosts({channelId: props.channel.id, beforeId: newestMessageId})); // eslint-disable-line no-await-in-loop
+                newestMessageId = data.order[data.order.length - 1];
+            } while (!data.posts[oldestPostId]);
```

Comment:
This might break/leak for a reason, maybe during no network connection or server error.